react_app/src/actions.js

```diff
@@ -0,0 +1,82 @@
+import * as ActionTypes from "./actionTypes";
+
+function checkAuthenticationSuccess(response) {
+  return { type: ActionTypes.CHECK_AUTHENTICATION_SUCCESS, response };
+}
+
+function checkAuthenticationFailed(error) {
+  return { type: ActionTypes.CHECK_AUTHENTICATION_FAILED, error };
+}
+
+function loginSuccess(response) {
+  return { type: ActionTypes.LOGIN_SUCCESS, response };
+}
+
+function loginFailed(error) {
+  return { type: ActionTypes.LOGIN_FAILED, error };
+}
+
+export function updateNewUser(newUser) {
+  return { type: ActionTypes.UPDATE_NEW_USER, payload: newUser };
+}
+
+export function login(formData) {
+  // Thunk
+  return async dispatch => {
+    dispatch({ type: "LOGIN" });
+    try {
+      fetch("/login", {
+        method: "POST",
+        credentials: "include",
+        body: formData
+      })
+        .then(response => response.json().then(body => ({ body, response })))
+        .then(({ body, response }) => {
+          if (response.status === 200) {
+            dispatch(loginSuccess(response));
+          } else {
+            dispatch(loginFailed(body.message));
+          }
+        })
+        .catch(error => {
+          dispatch(loginFailed("Unable to connect to server"));
+        });
+    } catch (err) {
+      dispatch(loginFailed(err.message));
+    }
+  };
+}
+
+export function logout() {
+  // Thunk
```

Comment:
Thunk?

Reply:
See how these action creators return functions rather than actions, and that function gets yielded the `dispatch` function? This is all handled by a piece of middleware called `redux-thunk`. It's not the only choice, but it is probably one of the most popular. I just whacked this comment in here out of habit to draw attention to `thunk` having been the choice here.

Reply:
Here look;
https://www.npmjs.com/package/redux-thunk

1.4m downloads a week. The other big alternative to putting side-effects like `fetch` into action creators is a thing called `redux-saga`, which is actually more powerful but thunk is just easier.